### ISSUE TITLE
[LORA] How to unload lora weight after load_lora_weights

### ISSUE THREAD

**Issue by yuananf** [huggingface/diffusers#3689]
**Is your feature request related to a problem? Please describe.**
I use pipe.load_lora_weights to load a lora weight, but don't know how to unload it.
I tried pipe.unet.set_default_attn_processor(), but it didn't work.

**Describe the solution you'd like**
I need a clear function to unload lora weight.

**Comment by patrickvonplaten** [comment]
cc @sayakpaul

**Comment by frankjoshua** [comment]
I'm looking for that same thing. The idea is to dynamically load Loras at different weights. Probably using prompts like <lora:filename:multiplier>.
Optimally we would like to preload multiple Loras and change their weights when doing inference.

**Comment by alexblattner** [comment]
you can do this with this: https://gist.github.com/takuma104/43552b8ec70b63323c57dc9c6fcb9b90#gistcomment-4563596

**Comment by sayakpaul** [comment]
> Probably using prompts like lora:filename:multiplier.

@frankjoshua

This design is atypical for `diffusers` since we prefer explicit exposition of arguments. @patrickvonplaten let me know if you think otherwise. 

FWIW, we do support passing the `scale` parameter as a part `cross_attention_kwargs`. The `scale` argument lets you control the weight with which the LoRA params will be "merged" with the base model params. Check out the doc to know more: https://huggingface.co/docs/diffusers/main/en/training/lora.

**Comment by frankjoshua** [comment]
> > Probably using prompts like lora:filename:multiplier.
> 
> @frankjoshua
> 
> This design is atypical for `diffusers` since we prefer explicit exposition of arguments. @patrickvonplaten let me know if you think otherwise.
> 
> FWIW, we do support passing the `scale` parameter as a part `cross_attention_kwargs`. The `scale` argument lets you control the weight with which the LoRA params will be "merged" with the base model params. Check out the doc to know more: https://huggingface.co/docs/diffusers/main/en/training/lora.

I think I was unclear. I'm not suggesting that the prompt is parsed by the diffuser. I would handle the prompt parsing. I just want to expose a way to control the LoRA weights during inference. Speed of inference is very important. Loading the model again with new weights is way too time consuming.

**Comment by sayakpaul** [comment]
> I would handle the prompt parsing. I just want to expose a way to control the LoRA weights during inference. Speed of inference is very important. Loading the model again with new weights is way too time consuming.

I think I answered the question on controlling `scale` of a LoRA checkpoint during inference. Could you shed more light on where you're facing slowdown? Preferably some code / Colab Notebook would be great for us to look further.

**Comment by sayakpaul** [comment]
> Personally I wanted to temporary suppress the LoRA not unload it completely. Maybe that's a separate issue though.

WDYM by temporarily suppressing LoRA? Maybe open a new issue for that?

**Comment by frankjoshua** [comment]
@patrickvonplaten provided a working sample of changing weights here: [](https://github.com/huggingface/diffusers/issues/3620#issuecomment-1581069756)
I tested it and it worked ok.
This issue maybe should be closed. I'm still having issues with LoRas running 50% slower and safetensors not loading with 16bit floats but that should be handled elsewhere.

**Comment by sayakpaul** [comment]
Sure, please feel free to create separate issue threads and we'd be happy to look into them. 

Closing the issue since I think the main issue discussed here is answered.

**Comment by yuananf** [comment]
> Coming to your questions, @yuananf. I ran some minor tests and here are my findings.
> 
> Let's load a dog LoRA first:
> 
> ```python
> import torch 
> from diffusers import StableDiffusionPipeline
> from PIL import Image
> 
> def image_grid(imgs, rows, cols):
>     assert len(imgs) == rows * cols
> 
>     w, h = imgs[0].size
>     grid = Image.new("RGB", size=(cols * w, rows * h))
>     grid_w, grid_h = grid.size
> 
>     for i, img in enumerate(imgs):
>         grid.paste(img, box=(i % cols * w, i // cols * h))
>     return grid
> 
> base_model_id = "runwayml/stable-diffusion-v1-5"
> lora_model_id = "patrickvonplaten/lora_dreambooth_dog_example"
> 
> pipeline = StableDiffusionPipeline.from_pretrained(
>     base_model_id, torch_dtype=torch.float16, safety_checker=None
> ).to("cuda")
> pipeline.load_lora_weights(lora_model_id)
> 
> prompt = "a photo of a sks dog"
> images = pipeline(
>     prompt=prompt, 
>     num_inference_steps=15, 
>     num_images_per_prompt=4,
>     generator=torch.manual_seed(0)
> ).images
> 
> image_grid(images, 1, 4)
> ```
> 
> This leads to:
> 
> ![image](https://user-images.githubusercontent.com/22957388/244353761-65477a1b-dfaa-4aa8-9a44-26968790fb66.png)
> 
> As we can see it can properly reflect the "sks" dog subject.
> 
> (DON'T worry about the quality as the number of steps and the scheduler choice is terrible)
> 
> Now if we do:
> 
> ```python
> from diffusers.models.attention_processor import AttnProcessor2_0
> 
> pipeline.unet.set_attn_processor(AttnProcessor2_0())
> 
> images = pipeline(
>     prompt=prompt, 
>     num_inference_steps=15, 
>     num_images_per_prompt=4,
>     generator=torch.manual_seed(0)
> ).images
> 
> image_grid(images, 1, 4)
> ```
> 
> Now, this leads to:
> 
> ![image](https://user-images.githubusercontent.com/22957388/244354129-283cafda-9942-4fe4-ba11-9914997b85bb.png)
> 
> (not reflecting my subject)
> 
> I guess this is what you're looking for?
> 
> This is my [Colab Notebook](https://colab.research.google.com/gist/sayakpaul/4ad63e98ca9838f659117d33ae086f6e/scratchpad.ipynb) where I ran my tests. It additionally has another A1111 LoRA checkpoint on which I conducted a similar test and my findings hold.
> 
> Cc: @patrickvonplaten

Thank you so much! This worked.

**Comment by pietrobolcato** [comment]
This method doesn't work for me. I think it has to do with the fact that we are only replacing back unet, and not the text encoder as well.

Here are some examples:

1. Generate without LoRA, a fresh loaded pipeline

![image](https://github.com/huggingface/diffusers/assets/3061306/929c9239-19f8-43a5-a123-7de8ef0220cc)


3. Generate with LoRA

![image](https://github.com/huggingface/diffusers/assets/3061306/c2381423-bf80-419c-84de-e073b475da16)


4. Remove LoRA:

```
self.text_to_image_pipeline.unet.set_attn_processor(AttnProcessor2_0())
```

![image](https://github.com/huggingface/diffusers/assets/3061306/fea3e348-c64d-4b40-9d23-c1412886a440)


All the settings are kept the same. It would be really great to have a built-in method in the pipeline to unload the LoRA weights, this is really helpful at inference time

**Comment by sayakpaul** [comment]
Introduced `unload_lora_weights()` in https://github.com/huggingface/diffusers/pull/4034.

**Comment by alkollo** [comment]
> > Probably using prompts like lora:filename:multiplier.
> 
> @frankjoshua
> 
> This design is atypical for `diffusers` since we prefer explicit exposition of arguments. @patrickvonplaten let me know if you think otherwise.
> 
> FWIW, we do support passing the `scale` parameter as a part `cross_attention_kwargs`. The `scale` argument lets you control the weight with which the LoRA params will be "merged" with the base model params. Check out the doc to know more: https://huggingface.co/docs/diffusers/main/en/training/lora.

This seems to lead to another issue, how do you set the scale for a defined Lora ?

I use this scheme:
```
pipe.load_lora_weights(loraA)
pipe.load_lora_weights(loraB)
 
image =  pipe.text2img(
    prompt=prompt,
    guidance_scale=guidance_scale,
    num_inference_steps=num_inference_steps,
    height=height,
    width=width,
    negative_prompt=forced_negative,
    generator=generator,
    max_embeddings_multiples=5,
    callback=lambda step, timestep, latents: progress_callback(
        step, timestep, latents, request_id, "progress"
    ),
    callback_steps=5,
    cross_attention_kwargs={"scale": 0.5},
).images[0]
```

This will apply 0.5 to which model ?
How to apply 0.5 to modelA and 0.2 to modelB for example ?

Based on the latest doc here : https://huggingface.co/docs/diffusers/main/en/training/lora

Thanks for your time.

**Comment by StateGovernment** [comment]
> Hey there, first of all thanks for the amazing work as always!
> 
> This method unfortunately doesn't work for me @sayakpaul. I think it has to do with the fact that we are only replacing back unet, and not the text encoder as well. I am using this [LoRA](https://civitai.com/models/60132/colored-icons-logo-generator) specifically.
> 
> Here are some examples:
> 
> 1. Generate without LoRA, a fresh loaded pipeline
> 
> ![image](https://user-images.githubusercontent.com/3061306/251840935-929c9239-19f8-43a5-a123-7de8ef0220cc.png)
> 
> 3. Generate with LoRA
> 
> ![image](https://user-images.githubusercontent.com/3061306/251840817-c2381423-bf80-419c-84de-e073b475da16.png)
> 
> 4. Remove LoRA:
> 
> ```
> self.text_to_image_pipeline.unet.set_attn_processor(AttnProcessor2_0())
> ```
> 
> ![image](https://user-images.githubusercontent.com/3061306/251841142-fea3e348-c64d-4b40-9d23-c1412886a440.png)
> 
> All the settings are kept the same. It would be really great to have a built-in method in the pipeline to unload the LoRA weights, this is really helpful at inference time!
> 
> If needed, this is the code whole code:
> 
> ```python
> """Text to image component class"""
> 
> import random
> import sys
> from typing import List
> 
> import torch
> from compel import Compel
> from diffusers import DiffusionPipeline, StableDiffusionPipeline
> from diffusers.models.attention_processor import AttnProcessor2_0
> from diffusers.pipelines.stable_diffusion import StableDiffusionPipelineOutput
> 
> from src.image_generation.components import utils
> from src.image_generation.components.postprocessor import Postprocessor
> 
> 
> class TextToImage:
>     """Initialize the text to image pipeline, and expose methods to perform inference"""
> 
>     def __init__(
>         self,
>         model_pipeline: DiffusionPipeline,
>         postprocessor: Postprocessor = None,
>         enable_xformers: bool = False,
>     ) -> None:
>         """Initalize the text to image pipeline
> 
>         Parameters
>         ----------
>         model_pipeline : DiffusionPipeline
>             The base model pipeline to copy the components from
>         postprocessor : Postprocessor, optional
>             The postprocessor class, initialized and with the model loaded
> 
>             Defaults to None
>         enable_xformers: bool, optional
>             If true, it enables xformers optimization for the pipelines. It needs the
>             package `xformer` to be installed for it to work.
> 
>             Defaults to False
>         """
>         self.device = utils.get_device()
>         self.postprocessor = postprocessor
> 
>         self.text_to_image_pipeline = StableDiffusionPipeline(
>             **model_pipeline.components,
>         ).to(self.device)
> 
>         self.compel_processor = Compel(
>             tokenizer=self.text_to_image_pipeline.tokenizer,
>             text_encoder=self.text_to_image_pipeline.text_encoder,
>         )
> 
>         if enable_xformers:
>             utils.enable_xformers(self.text_to_image_pipeline)
> 
>     def generate(
>         self,
>         prompt: str,
>         negative_prompt: str = "",
>         width: int = 512,
>         height: int = 512,
>         num_inference_steps: int = 20,
>         guidance_scale: float = 7.5,
>         num_images_per_prompt: int = 1,
>         seed: int = None,
>         apply_gfpgan: bool = False,
>         lora_path: str = None,
>     ) -> StableDiffusionPipelineOutput:
>         """Performs text to image inference
> 
>         Parameters
>         ----------
>         prompt : str
>             The prompt or prompts to guide the image generation. It supports prompt
>             weight using `+` and `-`, according to https://github.com/damian0815/compel
>         negative_prompt : str, optional
>             The prompt or prompts not to guide the image generation
> 
>             Defaults to ""
>         width : int, optional
>             The width of the generated image.
> 
>             Defaults to 512
>         height : int, optional
>             The height of the generated image.
> 
>             Defaults to 512
>         num_inference_steps : int, optional
>             The number of denoising steps. More denoising steps usually lead to a higher
>             quality image at the expense of slower inference.
> 
>             Defaults to 20
>         guidance_scale : float, optional
>             Higher guidance scale encourages to generate images that are closely linked
>             to the text prompt, usually at the expense of lower image quality.
> 
>             Defaults to 7.5
>         num_images_per_prompt : int, optional
>             The number of generated images per prompt
> 
>             Defaults to 1
>         seed : int, optional
>             An integer used as random seed to have reproducible results
>             If None, random seeds will be generated.
> 
>             Defaults to None
>         apply_gfpgan : bool, optional
>             If true, applies GPFGAN on the generated images. GFPGAN must be initialized
>             in the constructor, otherwise a warning is raised.
> 
>             Defaults to False
>         lora_path : str, optional
>             If specified, it loads a LoRA weight before running inference
> 
>             Defaults to None
> 
>         Returns
>         -------
>         StableDiffusionPipelineOutput:
>             An object containing the list of generated images, and a flag expressing if
>             NSFW was detected.
> 
>             The images can be accessed using the `images` property, and the NSFW flag
>             using the `nsfw_content_detected` property.
>         """
>         if seed is None:  # set random seed
>             seed = random.randint(0, sys.maxsize)
>         
>         if lora_path is not None:  # load LoRA
>             self.text_to_image_pipeline.load_lora_weights(lora_path)
>         else: # unload LoRA
>             self.text_to_image_pipeline.unet.set_attn_processor(AttnProcessor2_0())
> 
>         result = self.text_to_image_pipeline(
>             prompt_embeds=self.compel_processor(prompt),
>             negative_prompt_embeds=self.compel_processor(negative_prompt),
>             width=width,
>             height=height,
>             num_inference_steps=num_inference_steps,
>             guidance_scale=guidance_scale,
>             num_images_per_prompt=num_images_per_prompt,
>             generator=torch.manual_seed(seed),
>         )
> 
>         # apply postprocessing if model is initialized and flag set to true
>         if apply_gfpgan and self.postprocessor:
>             result.images = self.postprocessor.apply_gfpgan(result)
> 
>         return result
> ```

Thank you so much, 
`
self.text_to_image_pipeline.unet.set_attn_processor(AttnProcessor2_0())
`
does this completely remove Lora weights to a point where we can load another Lora without affecting anything?

**Comment by bigmover** [comment]
> @patrickvonplaten provided a working sample of changing weights here: I tested it and it worked ok. This issue maybe should be closed. I'm still having issues with LoRas running 50% slower and safetensors not loading with 16bit floats but that should be handled elsewhere.

Is there an example for dynamic loading/uploading lora weight？Would you mind to share it?

**Comment by crapthings** [comment]
> Introduced `unload_lora_weights()` in #4034.

how do i know if lora is already fused or not

**Comment by Aktsvigun** [comment]
Hello! Any updates here? @sayakpaul

**Comment by sayakpaul** [comment]
Call `unload_lora_weights()`
